backend/scripts/test_citation_builder.py

```diff
@@ -0,0 +1,281 @@
+            expected_text="""Growth! [[0]](https://0.com).""",
+            expected_citations=["doc_0"],
+        )
+        self.run_test_case(test_case)
+
+
+    def test_multiple_unique_citations(self):
+        test_case = SimpleCitationExtractionTestCase(
+            test_name="Multiple unique citations",
+            input_text=""""Test! [1] [5].""",
+            expected_text=""""Test! [[0]](https://0.com) [[2]](https://2.com).""",
+            expected_citations=["doc_0", "doc_2"],
+        )
+        self.run_test_case(test_case)
+
+    def test_repeated_citations(self):
+        test_case = SimpleCitationExtractionTestCase(
+            test_name="Repeated citations",
+            input_text="""Test! [1][1][1]. And some more [1][2].""",
+            expected_text="""Test! [[0]](https://0.com). And some more [[0]](https://0.com).""",
+            expected_citations=["doc_0"],
+        )
+        self.run_test_case(test_case)
+
+    def test_citations_at_sentence_boundaries(self):
+        test_case = SimpleCitationExtractionTestCase(
+            test_name="Citations at sentence boundaries",
+            input_text="""Citation at the end of a sentence.[2] Another sentence.[4]""",
+            expected_text="""Citation at the end of a sentence.[[0]](https://0.com) Another sentence.[[1]]()""",
+            expected_citations=["doc_0", "doc_1"],
+        )
+        self.run_test_case(test_case)
+
+    def test_citations_at_beginning_middle_and_end(self):
+        test_case = SimpleCitationExtractionTestCase(
+            test_name="Citations at beginning, middle, and end",
+            input_text="""[1] Citation at the beginning. [3] In the middle. At the end [5].""",
+            expected_text="""[[0]](https://0.com) Citation at the beginning. [[1]]() In the middle. At the end [[2]](https://2.com).""",
+            expected_citations=["doc_0", "doc_1", "doc_2"],
+        )
+        self.run_test_case(test_case)
+
+    def test_mixed_valid_and_invalid_citations(self):
+        test_case = SimpleCitationExtractionTestCase(
+            test_name="Mixed valid and invalid citations",
+            input_text="""Mixed valid and invalid citations [1][99][3][100][5].""",
+            expected_text="""Mixed valid and invalid citations [[0]](https://0.com)[99][[1]]()[100][[2]](https://2.com).""",
+            expected_citations=["doc_0", "doc_1", "doc_2"],
+        )
+        self.run_test_case(test_case)
+
+    def test_hard_test_consecutive(self):
+        # while True:
+        test_case_4 = SimpleCitationExtractionTestCase(
+            test_name="Hardest!",
+            input_text="""Multiple citations in one sentence [1][4][5]. """,
+            expected_text="""Multiple citations in one sentence [[0]](https://0.com)[[1]]()[[2]](https://2.com).""",
+            expected_citations=[("doc_0"), ("doc_1"), ("doc_2")],
+        )
+        test_case_4.run_test((mock_docs, mock_doc_mapping))
+
+
+if __name__ == '__main__':
```

Comment:
Let's not run pytests using the __name__, pytests have a specific way of running certain classes etc. We should have the standard way and maybe just document it in a README